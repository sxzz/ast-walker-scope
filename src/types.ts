import type { ParserPlugin } from '@babel/parser'
import type { Node } from '@babel/types'

export interface ParseOptions {
  filename?: string
  parserPlugins?: ParserPlugin[]
}

export type Scope = Record<string, Node>

export interface WalkerContext {
  skip: () => void
  remove: () => void
  replace: (node: Node) => void
}

export interface HookContext extends WalkerContext {
  parent: Node
  key: string
  index: number

  scope: Scope
  scopes: Scope[]
  level: number
}

export interface WalkerHooks {
  enter?: (this: HookContext, node: Node) => void
  leave?: (this: HookContext, node: Node) => void
}
