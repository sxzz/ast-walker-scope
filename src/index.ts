import { parse } from '@babel/parser'
import { walk } from 'estree-walker'
import {
  extractIdentifiers,
  isFunctionType,
  walkFunctionParams,
} from './utils/babel'
import type {
  HookContext,
  ParseOptions,
  Scope,
  WalkerContext,
  WalkerHooks,
} from './types'
import type {
  Identifier,
  Node,
  Program,
  VariableDeclaration,
} from '@babel/types'
import type { ParserPlugin } from '@babel/parser'

const ADVANCED_SCOPE: Node['type'][] = [
  'CatchClause',
  'ForInStatement',
  'ForOfStatement',
]

export const walkSource = (
  code: string,
  walkHooks: WalkerHooks,
  { filename, parserPlugins }: ParseOptions = {}
) => {
  const plugins: ParserPlugin[] = parserPlugins || []
  if (filename) {
    if (/\.tsx?$/.test(filename)) plugins.push('typescript')
    if (filename.endsWith('x')) plugins.push('jsx')
  }

  const ast = parse(code, {
    sourceType: 'module',
    plugins,
  })

  walkAST(ast.program, walkHooks)

  return { ast }
}

export const walkAST = (ast: Program, { enter, leave }: WalkerHooks) => {
  let currentScope: Scope = {}
  const scopeStack: Scope[] = [currentScope]

  walk(ast, {
    enter(this: WalkerContext, node: Node, parent, ...args) {
      enterNode(node, parent)
      enter?.call(getHookContext(this, [parent, ...args]), node)
    },

    leave(this: WalkerContext, node, parent, ...args) {
      leaveNode(node, parent)

      leave?.call(getHookContext(this, [parent, ...args]), node)
    },
  })

  function getHookContext(
    ctx: WalkerContext,
    [parent, key, index]: [Node, string, number]
  ): HookContext {
    const scope = scopeStack.reduce((prev, curr) => ({ ...prev, ...curr }), {})
    return {
      ...ctx,
      parent,
      key,
      index,

      scope,
      scopes: scopeStack,
      level: scopeStack.length,
    }
  }

  function enterNode(node: Node, parent: Node) {
    if (isFunctionType(node)) {
      scopeStack.push((currentScope = {}))
      walkFunctionParams(node, registerBinding)
    } else if (node.type === 'CatchClause') {
      // catch param
      scopeStack.push((currentScope = {}))
      if (node.param && node.param.type === 'Identifier')
        registerBinding(node.param)
      return
    } else if (
      node.type === 'ForOfStatement' ||
      node.type === 'ForInStatement'
    ) {
      scopeStack.push((currentScope = {}))
    } else if (
      node.type === 'BlockStatement' &&
      !isFunctionType(parent) &&
      !ADVANCED_SCOPE.includes(parent.type)
    ) {
      scopeStack.push((currentScope = {}))
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  function leaveNode(node: Node, _parent: Node) {
    if (node.type === 'BlockStatement') {
      scopeStack.pop()
      currentScope = scopeStack[scopeStack.length - 1]
    } else if (
      node.type === 'FunctionDeclaration' ||
      node.type === 'ClassDeclaration'
    ) {
      if (node.declare || !node.id) return
      registerBinding(node.id)
    } else if (node.type === 'VariableDeclaration') {
      walkVariableDeclaration(node)
    } else if (
      node.type === 'ExportNamedDeclaration' &&
      node.declaration &&
      node.declaration.type === 'VariableDeclaration'
    ) {
      walkVariableDeclaration(node.declaration)
    }
  }

  function walkVariableDeclaration(stmt: VariableDeclaration) {
    if (stmt.declare) return

    for (const decl of stmt.declarations) {
      for (const id of extractIdentifiers(decl.id)) {
        registerBinding(id)
      }
    }
  }

  function registerBinding(id: Identifier) {
    if (currentScope) {
      currentScope[id.name] = id
    } else {
      error(
        'registerBinding called without active scope, something is wrong.',
        id
      )
    }
  }

  function error(msg: string, node: Node) {
    const e = new Error(msg)
    ;(e as any).node = node
    throw e
  }
}
