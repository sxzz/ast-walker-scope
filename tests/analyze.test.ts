import { describe, expect, test } from 'vitest'
import {
  isCallExpression,
  isDeclaration,
  isExportNamedDeclaration,
} from '@babel/types'
import MagicString from 'magic-string'
import { analyze } from '../src'
import type { Node } from '@babel/types'
import type { HookContext, Scope, WalkHooks } from '../src'

function stringifyScope(scope: Scope) {
  return `{\n   | > ${Object.entries(scope)
    .map(
      ([key, node]) =>
        `${key}: ${
          node?.loc?.start
            ? `${node?.loc?.start.line}, ${node?.loc?.start.column}`
            : undefined
        }`
    )
    .join('\n   | > ')}\n   | }`
}

function prependLineNumber(code: string, s: MagicString) {
  let idx = 0
  for (const [lineNumber, line] of code.split('\n').entries()) {
    s.prependLeft(idx, `${String(lineNumber + 1).padStart(2)} | `)
    idx += line.length + 1
  }
}

function output(s: MagicString, node: Node, ctx: HookContext) {
  s.appendLeft(
    node.end!,
    `/* LEVEL: ${ctx.scopes.length} \n   | > ${stringifyScope(ctx.scope)} */`
  )
}

describe('analyze', async () => {
  const fixtures = import.meta.glob('./fixtures/*.{ts,js}', {
    eager: true,
    as: 'raw',
  })

  for (const [filename, content] of Object.entries(fixtures)) {
    test(filename, () => {
      const s = new MagicString(content)
      prependLineNumber(content, s)

      const hooks: WalkHooks = {
        leave(node) {
          if (
            (isDeclaration(node) || isCallExpression(node)) &&
            !isExportNamedDeclaration(node)
          ) {
            output(s, node, this)
          }

          // if (node.type === 'BlockStatement' && !isFunctionType(this.parent))
          //   output(s, node, this)
          // else if (
          //   isFunctionType(node) &&
          //   this.parent.type !== 'VariableDeclarator'
          // )
          //   output(s, node, this)
        },
      }
      analyze(content, hooks, { filename })

      expect(s.toString()).toMatchSnapshot()
    })
  }
})
