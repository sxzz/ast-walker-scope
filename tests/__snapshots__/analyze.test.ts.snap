// Vitest Snapshot v1

exports[`analyze > ./fixtures/basic.ts 1`] = `
" 1 | const level = 'root'/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | } */
 2 | export const topVar = 'foo'/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > topVar: 2, 13
   | } */
 3 | 
 4 | class topClass {}/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > topVar: 2, 13
   | > topClass: 4, 6
   | } */
 5 | 
 6 | function topFn() {}/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | } */
 7 | 
 8 | export const topFoo = () => {
 9 |   console.log(level)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | } */
10 | }/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | > topFoo: 8, 13
   | } */
11 | 
12 | {
13 |   const level = '2'/* LEVEL: 2 
   | > {
   | > level: 13, 8
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | > topFoo: 8, 13
   | } */
14 |   class Level2Class {
15 |     //
16 |   }/* LEVEL: 2 
   | > {
   | > level: 13, 8
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | > topFoo: 8, 13
   | > Level2Class: 14, 8
   | } */
17 | 
18 |   {
19 |     const level = '3'/* LEVEL: 3 
   | > {
   | > level: 19, 10
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | > topFoo: 8, 13
   | > Level2Class: 14, 8
   | } */
20 |     function Level3Fn() {
21 |       //
22 |     }/* LEVEL: 3 
   | > {
   | > level: 19, 10
   | > topVar: 2, 13
   | > topClass: 4, 6
   | > topFn: 6, 9
   | > topFoo: 8, 13
   | > Level2Class: 14, 8
   | > Level3Fn: 20, 13
   | } */
23 |   }
24 | 
25 |   // Level 2
26 | }
27 | 
28 | // Level 1
29 | "
`;

exports[`analyze > ./fixtures/catch-params.ts 1`] = `
" 1 | const level = 'root'/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | } */
 2 | try {
 3 |   const tryVar = '1'/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > tryVar: 3, 8
   | } */
 4 | } catch (err) {
 5 |   console.log(err)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > err: 4, 9
   | } */
 6 | }
 7 | 
 8 | console.log(level)/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | } */
 9 | "
`;

exports[`analyze > ./fixtures/destructuring-var.ts 1`] = `
" 1 | const {
 2 |   a,
 3 |   b,
 4 |   c: c1,
 5 |   d: { e },
 6 | } = {
 7 |   a: 'a',
 8 |   b: 'b',
 9 |   c: 'c',
10 |   d: { e: 'e' },
11 | }/* LEVEL: 1 
   | > {
   | > a: 2, 2
   | > b: 3, 2
   | > c1: 4, 5
   | > e: 5, 7
   | } */
12 | 
13 | const [a1, a2, , [a4]] = [1, 2, 3, [4]]/* LEVEL: 1 
   | > {
   | > a: 2, 2
   | > b: 3, 2
   | > c1: 4, 5
   | > e: 5, 7
   | > a1: 13, 7
   | > a2: 13, 11
   | > a4: 13, 18
   | } */
14 | "
`;

exports[`analyze > ./fixtures/fn-params.ts 1`] = `
" 1 | let level = 'root'/* LEVEL: 1 
   | > {
   | > level: 1, 4
   | } */
 2 | 
 3 | const foo = function (a, b) {
 4 |   console.log(a, b)/* LEVEL: 2 
   | > {
   | > level: 1, 4
   | > a: 3, 22
   | > b: 3, 25
   | } */
 5 | }/* LEVEL: 1 
   | > {
   | > level: 1, 4
   | > foo: 3, 6
   | } */
 6 | "
`;

exports[`analyze > ./fixtures/loop.js 1`] = `
" 1 | const level = 'root'/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | } */
 2 | 
 3 | for (const iterator/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > iterator: 3, 11
   | } */ of object) {
 4 |   console.log(iterator)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > iterator: 3, 11
   | } */
 5 | }
 6 | 
 7 | for (const key/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > key: 7, 11
   | } */ in object) {
 8 |   console.log(key)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > key: 7, 11
   | } */
 9 | }
10 | 
11 | for (let i = 0/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | } */; i < array.length; i++) {
12 |   console.log(i)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | } */
13 | }
14 | 
15 | while (1) {
16 |   console.log(level)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | } */
17 | }
18 | 
19 | do {
20 |   console.log(level)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | } */
21 | } while (1)
22 | ;[].forEach((item, idx) => {
23 |   console.log(item, idx)/* LEVEL: 2 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | > item: 22, 13
   | > idx: 22, 19
   | } */
24 | })/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | } */
25 | 
26 | console.log(level)/* LEVEL: 1 
   | > {
   | > level: 1, 6
   | > i: 11, 9
   | } */
27 | "
`;
